packages/docusaurus-mdx-loader/src/remark/toc/index.ts

```diff
@@ -124,9 +178,21 @@ export default plugin;
 async function createExportNode(name: string, object: any): Promise<MdxjsEsm> {
   const {valueToEstree} = await import('estree-util-value-to-estree');
 
+  const tocObject = object.map((heading: TOCItem | string) => {
+    if (typeof heading === 'string') {
+      const argumentName = heading.replace('...', '');
+      return {
+        type: 'SpreadElement',
+        argument: {type: 'Identifier', name: argumentName},
+      };
+    }
+
+    return valueToEstree(heading);
+  });
+
   return {
     type: 'mdxjsEsm',
-    value: `export const ${name} = ${stringifyObject(object)}`,
+    value: `export const ${name} = ${constructArrayString(object)}`,
```

Comment:
Little question for @wooorm or @dimaMachina

Could you explain when to use `mdxjsEsm.value` and when to use `mdxjsEsm.data.estree.body`? 

It's not clear to me which one will be used for what exactly.

On the Nextra PR, it seems they don't use `mdxjsEsm.value` and yet it works. However if I remove it on our remark plugin the toc is not exported anymore.

---

Original Nextra code:

https://github.com/shuding/nextra/pull/2199/files

<img width="683" alt="CleanShot 2024-01-12 at 22 40 56@2x" src="https://github.com/facebook/docusaurus/assets/749374/5e1f8262-77b2-4da5-a13f-e00cd238f4fe">

Reply:
It’s the same thing as when strings of raw HTML are supported and when ASTs of nodes are supported. The string is used by tools that don’t understand JavaScript ASTs (remark). The AST is used by tools that don’t understand strings of JavaScript (recma).

In this case, you are not generating MDX files—you are not compiling to markdown. So you don’t need `value`.
You are generating JavaScript. So you need JavaScript ASTs.

> However if I remove it on our remark plugin the toc is not exported anymore.

I don’t know what these words practically mean. You are explaining a bug that I very much doubt would happen.

---

> [src/remark/toc/utils.ts](https://github.com/facebook/docusaurus/pull/9684#diff-50ad45314b03426612bb4a598f8adf0cc099f4bfc8674528aa358ed01aaaacaf)

I recommend dropping `isImport`, `isExport`, and `hasExports`. The types aren’t very good, as you can see from the casts. Take a look at the places these utilities are used, and which types TypeScript knows of there. TS is already much smarter there.

Reply:
From my testing, removing value doesn't break the generation of ToC. But I'm concerned with the possibility of another plugin relying on `value`. So I feel like it would be right to keep it. At least since it's a required property in `MdxjsEsm`.

Reply:
I recommend an empty string `''` for now instead of pulling in `astring`. Practically there are no plugins that do, or should, use the string of JavaScript, in turning markdown into javascript.
You can always add it later when a plugin like that appears!

Reply:
Thanks for the answer @wooorm I understand better now.

The problem was probably only in our unit test snapshots since we serialize back to Markdown:

```tsx
const processFixture = async (name: string) => {
  const {remark} = await import('remark');
  const {default: gfm} = await import('remark-gfm');
  const {default: mdx} = await import('remark-mdx');

  const filePath = path.join(__dirname, '__fixtures__', `${name}.md`);
  const file = await vfile.read(filePath);
  const result = await remark()
    .use(headings)
    .use(gfm)
    .use(mdx)
    .use(plugin)
    .process(file);

  return result.value;
};
```

This makes the snapshots more convenient to review than using AST or JSX IMHO, so I'd like to keep this if possible.

But as far as I understand it keeping this format would require using astring?

Do you see any solution to this problem? 

I was thinking we could create an astring remark plugin generating values just during tests for example?

Reply:
Can confirm it works nicely and helps keeping readable snapshots:

```js
visit(root, (child) => {
        if (child.type === 'mdxjsEsm') {
          const node = child as MdxjsEsm;
          node.value = node.data?.estree
            ? generate(node.data.estree)
            : node.value;
        }
});
```

<img width="495" alt="CleanShot 2024-01-18 at 18 13 02@2x" src="https://github.com/facebook/docusaurus/assets/749374/f8a9834c-864e-4824-975e-8dffc1d49a3c">

Reply:
> The problem was probably only in our unit test snapshots since we serialize back to Markdown:

Here you specifically care about JavaScript; not markdown.
So use MDX to compile to JavaScript. Snapshot that. Much easier to review than a fake string that never actually runs!

`compile(input, {remarkPlugins, rehypePlugins})`

Reply:
👍 makes sense, will change the snapshot format

Reply:
btw don’t treat my comments as blockers — just random ideas about how I would do it!